File: src/rchain/chat_models.rs
use std::env;
use std::error::Error;

use reqwest::blocking::Client;
use serde_json::{json, Map, Value};

use crate::rchain::ai::AIMessage;
use crate::rchain::human::HumanMessage;
use crate::rchain::tools::{ToolCall, ToolDefinition};

/// Fireworks chat-completions client.
#[derive(Debug, Clone)]
pub struct ChatFireworks {
    model: String,
    temperature: f64,
    api_key: String,
    base_url: String,
    client: Client,
    tools: Option<Vec<ToolDefinition>>,
}

impl ChatFireworks {
    /// Creates a new client from model id and `FIREWORKS_API_KEY` env var.
    pub fn new(model: impl Into<String>, temperature: f64) -> Result<Self, Box<dyn Error>> {
        let api_key = env::var("FIREWORKS_API_KEY")
            .map_err(|_| "FIREWORKS_API_KEY is not set in the environment")?;
        Ok(Self {
            model: model.into(),
            temperature,
            api_key,
            base_url: "https://api.fireworks.ai/inference/v1/chat/completions".to_string(),
            client: Client::new(),
            tools: None,
        })
    }

    /// Returns a cloned client bound to tool definitions.
    pub fn bind_tools(&self, tools: Vec<ToolDefinition>) -> Self {
        let mut bound = self.clone();
        bound.tools = Some(tools);
        bound
    }

    /// Invokes the model with user messages only.
    pub fn invoke(&self, messages: &[HumanMessage]) -> Result<AIMessage, Box<dyn Error>> {
        let chat_messages = messages
            .iter()
            .map(|message| ChatMessage::user(message.clone()))
            .collect::<Vec<_>>();
        self.invoke_messages(&chat_messages)
    }

    /// Invokes the model with fully-typed role messages.
    pub fn invoke_messages(&self, messages: &[ChatMessage]) -> Result<AIMessage, Box<dyn Error>> {
        let mut payload = Map::new();
        payload.insert("model".to_string(), Value::String(self.model.clone()));
        payload.insert(
            "messages".to_string(),
            Value::Array(messages.iter().map(|message| message.to_json()).collect()),
        );
        payload.insert("temperature".to_string(), json!(self.temperature));
        if let Some(tools) = &self.tools {
            payload.insert(
                "tools".to_string(),
                Value::Array(tools.iter().map(|tool| tool.to_json()).collect()),
            );
        }
        let payload = Value::Object(payload);

        let response = self
            .client
            .post(&self.base_url)
            .bearer_auth(&self.api_key)
            .json(&payload)
            .send()?;

        if !response.status().is_success() {
            let status = response.status();
            let body = response.text().unwrap_or_default();
            return Err(format!("Fireworks API error {status}: {body}").into());
        }

        let body: serde_json::Value = response.json()?;
        let message = &body["choices"][0]["message"];
        let content = message["content"].as_str().unwrap_or("").to_string();
        let tool_calls = parse_tool_calls(message);

        Ok(AIMessage {
            content,
            tool_calls,
        })
    }
}

/// Supported role values in chat requests.
#[derive(Debug, Clone)]
pub enum MessageRole {
    /// Human/user role.
    User,
    /// Assistant role.
    Assistant,
    /// Tool result role.
    Tool,
}

/// Chat message wrapper used by [`ChatFireworks`].
#[derive(Debug, Clone)]
pub struct ChatMessage {
    role: MessageRole,
    content: Value,
    tool_call_id: Option<String>,
    tool_calls: Option<Vec<ToolCall>>,
}

impl ChatMessage {
    /// Builds a user message from [`HumanMessage`].
    pub fn user(message: HumanMessage) -> Self {
        Self {
            role: MessageRole::User,
            content: message.to_json(),
            tool_call_id: None,
            tool_calls: None,
        }
    }

    /// Builds a plain-text user message.
    pub fn user_text(content: impl Into<String>) -> Self {
        Self {
            role: MessageRole::User,
            content: Value::String(content.into()),
            tool_call_id: None,
            tool_calls: None,
        }
    }

    /// Builds a multipart user message.
    pub fn user_parts(parts: Vec<Value>) -> Self {
        Self {
            role: MessageRole::User,
            content: Value::Array(parts),
            tool_call_id: None,
            tool_calls: None,
        }
    }

    /// Builds an assistant message from an [`AIMessage`].
    pub fn assistant_from_ai(message: &AIMessage) -> Self {
        let content = if message.content.is_empty() {
            Value::Null
        } else {
            Value::String(message.content.clone())
        };
        Self {
            role: MessageRole::Assistant,
            content,
            tool_call_id: None,
            tool_calls: if message.tool_calls.is_empty() {
                None
            } else {
                Some(message.tool_calls.clone())
            },
        }
    }

    /// Builds a tool-result message associated with a tool call id.
    pub fn tool_result(tool_call_id: impl Into<String>, content: impl Into<String>) -> Self {
        Self {
            role: MessageRole::Tool,
            content: Value::String(content.into()),
            tool_call_id: Some(tool_call_id.into()),
            tool_calls: None,
        }
    }

    /// Serializes this chat message to provider JSON format.
    pub fn to_json(&self) -> Value {
        let mut map = Map::new();
        map.insert(
            "role".to_string(),
            Value::String(
                match self.role {
                    MessageRole::User => "user",
                    MessageRole::Assistant => "assistant",
                    MessageRole::Tool => "tool",
                }
                .to_string(),
            ),
        );
        map.insert("content".to_string(), self.content.clone());
        if let Some(tool_call_id) = &self.tool_call_id {
            map.insert(
                "tool_call_id".to_string(),
                Value::String(tool_call_id.clone()),
            );
        }
        if let Some(tool_calls) = &self.tool_calls {
            map.insert(
                "tool_calls".to_string(),
                Value::Array(tool_calls.iter().map(|call| call.to_json()).collect()),
            );
        }
        Value::Object(map)
    }
}

fn parse_tool_calls(message: &Value) -> Vec<ToolCall> {
    let mut tool_calls = Vec::new();
    if let Some(calls) = message["tool_calls"].as_array() {
        for call in calls {
            let id = call["id"].as_str().unwrap_or("").to_string();
            let name = call["function"]["name"].as_str().unwrap_or("").to_string();
            let arguments = &call["function"]["arguments"];
            let args = match arguments {
                Value::String(raw) => {
                    serde_json::from_str(raw).unwrap_or(Value::String(raw.clone()))
                }
                other => other.clone(),
            };
            if !name.is_empty() {
                tool_calls.push(ToolCall { id, name, args });
            }
        }
    }
    tool_calls
}

File: mpipe/rchain/chat_models.py
# SPDX-License-Identifier: GPL-3.0-or-later
#
# mpipe - Multi-provider LLM CLI tools
# Copyright (C) 2026 Nicolas Schaetti and contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""mpipe.rchain.chat_models module.

Notes
-----
This module is part of the Python port of the `mpipe` project.
"""
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from enum import Enum
from typing import Any

import requests

from mpipe.rchain.ai import AIMessage
from mpipe.rchain.human import HumanMessage
from mpipe.rchain.tools import ToolCall, ToolDefinition


class MessageRole(str, Enum):
    """Messagerole.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"
# end class MessageRole


@dataclass(slots=True)
class ChatMessage:
    """Chatmessage.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    role: MessageRole
    content: Any
    tool_call_id: str | None = None
    tool_calls: list[ToolCall] | None = None

    @classmethod
    def user(cls, message: HumanMessage) -> "ChatMessage":
        """User.

        Parameters
        ----------
        message : HumanMessage
            Argument value.

        Returns
        -------
        'ChatMessage'
            Returned value.
        """
        return cls(role=MessageRole.USER, content=message.to_json())
    # end def user

    @classmethod
    def user_text(cls, content: str) -> "ChatMessage":
        """User text.

        Parameters
        ----------
        content : str
            Argument value.

        Returns
        -------
        'ChatMessage'
            Returned value.
        """
        return cls(role=MessageRole.USER, content=content)
    # end def user_text

    @classmethod
    def user_parts(cls, parts: list[dict[str, Any]]) -> "ChatMessage":
        """User parts.

        Parameters
        ----------
        parts : list[dict[str, Any]]
            Argument value.

        Returns
        -------
        'ChatMessage'
            Returned value.
        """
        return cls(role=MessageRole.USER, content=parts)
    # end def user_parts

    @classmethod
    def assistant_from_ai(cls, message: AIMessage) -> "ChatMessage":
        """Assistant from ai.

        Parameters
        ----------
        message : AIMessage
            Argument value.

        Returns
        -------
        'ChatMessage'
            Returned value.
        """
        content = None if message.content == "" else message.content
        calls = message.tool_calls if message.tool_calls else None
        return cls(role=MessageRole.ASSISTANT, content=content, tool_calls=calls)
    # end def assistant_from_ai

    @classmethod
    def tool_result(cls, tool_call_id: str, content: str) -> "ChatMessage":
        """Tool result.

        Parameters
        ----------
        tool_call_id : str
            Argument value.
        content : str
            Argument value.

        Returns
        -------
        'ChatMessage'
            Returned value.
        """
        return cls(role=MessageRole.TOOL, content=content, tool_call_id=tool_call_id)
    # end def tool_result

    def to_json(self) -> dict[str, Any]:
        """To json.

        Parameters
        ----------
        None
            This callable does not accept explicit parameters.

        Returns
        -------
        dict[str, Any]
            Returned value.
        """
        payload: dict[str, Any] = {"role": self.role.value, "content": self.content}
        if self.tool_call_id:
            payload["tool_call_id"] = self.tool_call_id
        # end if
        if self.tool_calls:
            payload["tool_calls"] = [call.to_json() for call in self.tool_calls]
        # end if
        return payload
    # end def to_json
# end class ChatMessage


@dataclass(slots=True)
class ChatFireworks:
    """Chatfireworks.

    Notes
    -----
    This class follows the same role and hierarchy as the Rust implementation.
    """
    model: str
    temperature: float
    api_key: str
    base_url: str = "https://api.fireworks.ai/inference/v1/chat/completions"
    tools: list[ToolDefinition] | None = None

    @classmethod
    def new(cls, model: str, temperature: float) -> "ChatFireworks":
        """New.

        Parameters
        ----------
        model : str
            Argument value.
        temperature : float
            Argument value.

        Returns
        -------
        'ChatFireworks'
            Returned value.
        """
        api_key = os.getenv("FIREWORKS_API_KEY", "")
        if not api_key:
            raise ValueError("FIREWORKS_API_KEY is not set in the environment")
        # end if
        return cls(model=model, temperature=temperature, api_key=api_key)
    # end def new

    def bind_tools(self, tools: list[ToolDefinition]) -> "ChatFireworks":
        """Bind tools.

        Parameters
        ----------
        tools : list[ToolDefinition]
            Argument value.

        Returns
        -------
        'ChatFireworks'
            Returned value.
        """
        return ChatFireworks(
            model=self.model,
            temperature=self.temperature,
            api_key=self.api_key,
            base_url=self.base_url,
            tools=list(tools),
        )
    # end def bind_tools

    def invoke(self, messages: list[HumanMessage]) -> AIMessage:
        """Invoke.

        Parameters
        ----------
        messages : list[HumanMessage]
            Argument value.

        Returns
        -------
        AIMessage
            Returned value.
        """
        chat_messages = [ChatMessage.user(message) for message in messages]
        return self.invoke_messages(chat_messages)
    # end def invoke

    def invoke_messages(self, messages: list[ChatMessage]) -> AIMessage:
        """Invoke messages.

        Parameters
        ----------
        messages : list[ChatMessage]
            Argument value.

        Returns
        -------
        AIMessage
            Returned value.
        """
        payload: dict[str, Any] = {
            "model": self.model,
            "messages": [message.to_json() for message in messages],
            "temperature": self.temperature,
        }
        if self.tools is not None:
            payload["tools"] = [tool.to_json() for tool in self.tools]
        # end if

        response = requests.post(
            self.base_url,
            headers={"Authorization": f"Bearer {self.api_key}"},
            json=payload,
            timeout=60,
        )
        if not response.ok:
            raise RuntimeError(f"Fireworks API error {response.status_code}: {response.text}")
        # end if

        body = response.json()
        message = body.get("choices", [{}])[0].get("message", {})
        content = message.get("content") or ""
        tool_calls = _parse_tool_calls(message)
        return AIMessage(content=content, tool_calls=tool_calls)
    # end def invoke_messages
# end class ChatFireworks


def _parse_tool_calls(message: dict[str, Any]) -> list[ToolCall]:
    """ parse tool calls.

    Parameters
    ----------
    message : dict[str, Any]
        Argument value.

    Returns
    -------
    list[ToolCall]
        Returned value.
    """
    tool_calls: list[ToolCall] = []
    calls = message.get("tool_calls")
    if not isinstance(calls, list):
        return tool_calls
    # end if

    for call in calls:
        if not isinstance(call, dict):
            continue
        # end if
        call_id = str(call.get("id", ""))
        function = call.get("function") if isinstance(call.get("function"), dict) else {}
        name = str(function.get("name", ""))
        raw_arguments = function.get("arguments", {})
        if isinstance(raw_arguments, str):
            try:
                args: Any = json.loads(raw_arguments)
            except Exception:
                args = raw_arguments
            # end try
        else:
            args = raw_arguments
        # end if
        if name:
            tool_calls.append(ToolCall(id=call_id, name=name, args=args))
        # end if
    # end for
    return tool_calls
# end def _parse_tool_calls
